acceptance/helpers.rb

```diff
@@ -66,173 +203,205 @@ def prepare_upgrade_with(initial_package_version_or_collection)
-        manifest_contents = %(node default { #{site_pp_contents} })
+      def apply_manifest_on_agents(manifest)
+        agent_nodes = agents_only.map do |agent|
+          "'#{agent.to_s}'"
+        end.join(', ')
+        site_pp_contents = %(node #{agent_nodes} { #{manifest} })
 
         # PMT will have installed dependencies in the production environment; We will put our manifest there, too:
         site_pp_path = File.join(puppet_config(master, 'codedir'), 'environments', 'production', 'manifests', 'site.pp')
 
-        step "Save current site.pp" do
+        cleanup = nil
+
+        step "(Master) Save the current site.pp file" do
           if file_exists_on(master, site_pp_path)
             original_contents = file_contents_on(master, site_pp_path)
             original_perms = on(master, %(stat -c "%a" #{site_pp_path})).stdout.strip
-
-            teardown do
-              step "restore original manifest" do
+  
+            cleanup = lambda do
+              step "(Master) Restore the original site.pp file" do
                 on(master, %(echo "#{original_contents}" > #{site_pp_path}))
                 on(master, %(chmod #{original_perms} #{site_pp_path}))
               end
             end
           else
-            teardown do
+            cleanup = lambda do
               on(master, "rm -f #{site_pp_path}")
             end
           end
         end
 
-        step "create site.pp on master with manifest:\n#{manifest_contents}" do
-          create_remote_file(master, site_pp_path, manifest_contents)
+        step "(Master) Create the new site.pp file with manifest:\n#{manifest}" do
+          create_remote_file(master, site_pp_path, manifest)
           on(master, %(chown #{puppet_user(master)} "#{site_pp_path}"))
           on(master, %(chmod 755 "#{site_pp_path}"))
         end
 
-        step "Execute puppet runs" do
-          with_puppet_running_on(master, master_opts) do
-            on(agents_only, puppet(%(agent --test --server #{master.hostname})), acceptable_exit_codes: [0, 2])
-            yield if block_given?
-          end
+        step "(Agents) Run Puppet to apply the manifest" do
+          on(agents_only, puppet(%(agent --test --server #{master.hostname})), acceptable_exit_codes: [0, 2])
         end
+      ensure
+        cleanup.call if cleanup
       end
 
-      # Wraps {with_default_site_pp} to:
+      # Generates a manifest that declares the puppet_agent class.
       #
-      # - Install the a puppet-agent package (`initial_package_version_or_collection`)
-      #   on all the agent hosts,
-      # - Put a teardown step in place so that this package is uninstalled after the test,
-      # - Run a default manifest on all the agent hosts, and
-      # - Allow for assertions inside in a block
+      # @param [Hash] params The class parameters
+      # @return [String] The generated puppet_agent class declaration
+      def class__puppet_agent(params = {})
```

Comment:
Nit: Do we lose any clarity around what the manifest should look like by including this helper?

In the original tests seeing the manifest in it's entirety might make it easier for someone new to the codebase to grasp what the manifest would look like.  Not a huge deal since the helper just wraps the manifest params in the classname, but if I were new to the module I would still need to look up this helper to find that out.